### ISSUE TITLE
Distinguish between C and C++ headers by keywords.

### ISSUE THREAD

**Issue by alexey-milovidov** [boyter/scc#213]
I see that `languages.json` supports `keywords` to distinguish between Coq and Verilog.
Let's apply it also for C and C++.

Keywords (and just more or less specific words) for C++:

namespace
template
using
class
constexpr
auto
bool
true
false
typename
throw
std
explicit
private
public
protected
virtual
override
default
noexcept
catch
delete
static_cast
dynamic_cast
reinterpret_cast
decltype

**Comment by justjosias** [comment]
Some of these can easily appear in C code though, so the detector has to be smart.

**Comment by alexey-milovidov** [comment]
That's true. I did not check the sources of scc, maybe it can use some threshold... the classification still be inaccurate but reasonably good heuristic should be Ok.

**Comment by boyter** [comment]
So yes that's what that logic is there for. It works by counting the keywords for each duplicate indicated by them sharing an extension and then seeing which one has the most keywords.

Im prepared to put this in if someone can help with the keywords, and ensuring that the false positive rate is really driven down to ideally 0%.

There is a workaround though, the --remap-all flag would allow this to happen right now, although it won't work as you would want for instances where terms are shared.

**Comment by alexey-milovidov** [comment]
> then seeing which one has the most keywords

It can be more difficult because there are no distinguishing (popular) keywords in C that are not used in C++.
We only limited to counting C++ keywords...

BTW, we have to filter comments, otherwise "private", "class", ... will be too common.

I've made an experiment with Linux codebase:

```
torvalds/linux$ find . -name '*.h' | while read file; do gcc -w -fpreprocessed -dD -E $file 2>/dev/null | grep -c --color -P '\b(namespace|template|using|class|constexpr|auto|typename|throw|std|explicit|private|public|protected|virtual|override|noexcept|catch|delete|static_cast|dynamic_cast|reinterpret_cast|decltype)\b'; done | awk '{ ++total; if ($1 > 0) { ++matches } } END { print total, matches }'
21284 482
```

I assume that it is 100% C (actually the assumption does not hold - there are some C++ files but not too much).

The false positive rate is 2.2%. The most popular found keywords are `class`, `private` and `delete`.
If we remove them:

```
torvalds/linux$ find . -name '*.h' | while read file; do gcc -w -fpreprocessed -dD -E $file 2>/dev/null | grep -c --color -P '\b(namespace|template|using|constexpr|auto|typename|throw|std|explicit|public|protected|virtual|override|noexcept|catch|static_cast|dynamic_cast|reinterpret_cast|decltype)\b'; done | awk '{ ++total; if ($1 > 0) { ++matches } } END { print total, matches }'
21284 158
```

0.7%

We need more experiments to validate these numbers on other codebases.

Maybe we can do something more sophisticated: instead of couning keywords count regex matching from the list of regexps. It can be done rather efficiently (100..1000 MB/sec/core on modern hardware) with Hyperscan.

**Comment by alexey-milovidov** [comment]
Fun fact:

I visited StackOverflow to find how to easily strip comments from C file:
https://stackoverflow.com/questions/2394017/remove-comments-from-c-c-code

The second answer is: use `scc`.
I thought that there is a command line option to `scc` to just strip comments from a file but did not find it.
But it's appeared to be different `scc`: https://github.com/jleffler/scc-snapshots

What a coincidence :rofl:

**Comment by boyter** [comment]
Yeah I had a feeling that C++ being a superset of C would make this... difficult. Although not impossible, if the count is the same it falls back to the first, which would be C in most cases, so even if there were shared keywords so long as they are in both it should even out on average.

As you mention I had a feeling the error rate would not be as close to 0 as I would like :) 

Adding regex... is probably going to be a non starter I think. Any regex engine is going to slow things down most likely, even if it was hyper scan which is admittedly very impressive.

Ill have a think on this a bit more. It might be worth code spelunking how cloc does this as it likely faces the same issue, and also just as likely has solved it.

Interesting to see there is another sci out there! I thought I was being reasonably unique with the name, but alas.

**Comment by boyter** [comment]
This should work now. The way linguist does this has been adopted and should work better.
